packages/firestore/src/api/user_data_converter.ts

```diff
@@ -297,19 +299,54 @@ export class UserDataConverter {
   }
 
   /** Parse document data from a set() call with '{merge:true}'. */
-  parseMergeData(methodName: string, input: AnyJs): ParsedSetData {
+  parseMergeData(
+    methodName: string,
+    input: AnyJs,
+    fieldPaths?: Array<string | firestore.FieldPath>
+  ): ParsedSetData {
     const context = new ParseContext(
       UserDataSource.MergeSet,
       methodName,
       FieldPath.EMPTY_PATH
     );
     validatePlainObject('Data must be an object, but it was:', context, input);
 
-    const updateData = this.parseData(input, context);
-    const fieldMask = new FieldMask(context.fieldMask);
+    const updateData = this.parseData(input, context) as ObjectValue;
+    let fieldMask: FieldMask | null = null;
+
+    if (!fieldPaths) {
+      fieldMask = new FieldMask(context.fieldMask);
+    } else {
+      const validatedFieldPaths = [];
+
+      for (const stringOrFieldPath of fieldPaths) {
+        let fieldPath;
+
+        if (stringOrFieldPath instanceof ExternalFieldPath) {
+          fieldPath = stringOrFieldPath as ExternalFieldPath;
+        } else if (typeof stringOrFieldPath === 'string') {
+          fieldPath = fieldPathFromDotSeparatedString(
+            methodName,
+            stringOrFieldPath
+          );
+        } else {
+          fail('Expected stringOrFieldPath to be a string or a FieldPath');
+        }
+
+        if (!updateData.contains(fieldPath)) {
+          throw new FirestoreError(
+            Code.INVALID_ARGUMENT,
+            `Field '${fieldPath}' is specified in your field mask but missing from your input data.`
+          );
+        }
+        validatedFieldPaths.push(fieldPath);
```

Comment:
What about overlapping field paths?

Reply:
We discussed this during API review at one point and decided to "do whatever Android" does. Android doesn't validate and uses the most broad field path if multiple field paths match.